iis-wf/iis-wf-referenceextraction/src/main/java/eu/dnetlib/iis/wf/referenceextraction/patent/PatentMetadataExtractorJob.java

```diff
@@ -0,0 +1,138 @@
+            JavaRDD<DocumentText> toBeProcessedContents = avroLoader.loadJavaRDD(sc, params.inputDocumentTextPath, DocumentText.class);
+            
+            JavaPairRDD<CharSequence, Tuple2<DocumentText, ImportedPatent>> pairedInput = toBeProcessedContents
+                    .mapToPair(x -> new Tuple2<CharSequence, DocumentText>(x.getId(), x))
+                    .join(importedPatent.mapToPair(x -> new Tuple2<CharSequence, ImportedPatent>(x.getApplnNr(), x)));
+            
+            JavaRDD<PatentOrFault> parsedPatents = pairedInput.map(x -> parse(x._2._1, x._2._2, parser));
+            parsedPatents.cache();
+            
+            Tuple2<JavaRDD<Patent>, JavaRDD<Fault>> splitPatentAndFault = splitPatentAndFault(parsedPatents);
+            
+            avroSaver.saveJavaRDD(splitPatentAndFault._1, Patent.SCHEMA$, params.outputPath);
+            avroSaver.saveJavaRDD(splitPatentAndFault._2, Fault.SCHEMA$, params.outputFaultPath);
+            avroSaver.saveJavaRDD(generateReportEntries(sc, splitPatentAndFault), ReportEntry.SCHEMA$,
+                    params.outputReportPath);
+        }
+    }
+
+    // ------------------------ PRIVATE --------------------------
+
+    private static JavaRDD<ReportEntry> generateReportEntries(JavaSparkContext sparkContext, 
+            Tuple2<JavaRDD<Patent>, JavaRDD<Fault>> splitPatentAndFault) {
+        
+        Preconditions.checkNotNull(sparkContext, "sparkContext has not been set");
+        
+        ReportEntry processedPatentsCounter = ReportEntryFactory.createCounterReportEntry(COUNTER_PROCESSED_TOTAL, splitPatentAndFault._1.count());
+        ReportEntry processedFaultsCounter = ReportEntryFactory.createCounterReportEntry(COUNTER_PROCESSED_FAULT, splitPatentAndFault._2.count());
+        
+        return sparkContext.parallelize(Lists.newArrayList(processedPatentsCounter, processedFaultsCounter));
+    }
+    
+    private static Tuple2<JavaRDD<Patent>, JavaRDD<Fault>> splitPatentAndFault(JavaRDD<PatentOrFault> parsedPatents) {
+        return new Tuple2<>(parsedPatents.map(e -> e.getPatent()), 
+                parsedPatents.filter(e -> e.getException() != null).map(
+                        e -> FaultUtils.exceptionToFault(e.getPatent().getApplnNr(), e.getException(), null)));
+    }
+    
+    private static PatentOrFault parse(DocumentText patent, ImportedPatent importedPatent, PatentMetadataParser parser) {
+        Patent.Builder resultBuilder = fillDataFromImport(Patent.newBuilder(), importedPatent);
+        try {
+            return new PatentOrFault(parser.parse(patent.getText(), resultBuilder).build());
+        } catch (PatentMetadataParserException e) {
+            log.error("error while parsing xml contents of patent id " + patent.getId() + ", text content: "
+                    + patent.getText(), e);
+            return new PatentOrFault(resultBuilder.build(), e);
```

Comment:
`PatentOrFault` name indicates an alternative field name, not sure why I stuck with `Exception` rather than `Fault`.

Anyway I decided to replace this class with simple `Tuple2<Patent, Fault>` which does exactly the same. I have also refactored the way `Fault` object is being built.

The thing is some of the Patent metadata come from importer and some from XML extraction. I decided to produce Patent object also when XML extraction fails to get the minimum set of Patent metadata we could match publication with. 
I am still scratching my head whether it is a good idea... It makes it possible to provide whatever metadata we could get (also when only a part of XML metadata could be retrieved) but on the other hand if our parser is incompatible with XML record maybe we should not work too hard to produce any crippled Patent object? Maybe we should simply return the fault without partially extracted patent metadata?

Reply:
Introducing `Tuple2` makes my comment invalid, but the original solution with `PatentOrFault` was very good and I liked it :)

Are we able to do anything useful with a crippled Patent ? If yes then the we should keep this solution, if not then we should produce only valid Patents or Faults.

Reply:
Well, it will always contain the meta from `ImportedPatent` which are mostly identifiers and it may be able to extract some of the metadata fields out of the XML record before it fails. So it depends on when it fails... If it manages to extract the title then maybe it is enough to become a valid patent entity otherwise we would need to discard an entity and all the matches between given patent and publications.

But lets just follow the KISS rule and not over-complicate things. I will make it discard the Patent whenever any kind of failure occurs.

Reply:
I can confirm: I have reimplemented this piece of error handling and `Patent` entity is not being returned anymore whenever any error occurs. This should be more straightforward.